Stabilize alloc::Layout const functions

ChrisDenton opened the pull request:
Stabilizes #67521. In particular the following stable methods are stabilized as `const fn`:

* `size`
* `align`
* `from_size_align`

Stabilizing `size` and `align` should not be controversial as they are simple (usize and NonZeroUsize) fields and I don't think there's any reason to make them not const compatible in the future. That being true, the other methods are trivially `const`. The only other issue being returning a `Result` from a `const fn` but this has been made more usable by recent stabilizations.

rust-highfive commented:
r? @m-ou-se

(rust_highfive has picked a reviewer for you, use r? to override)

m-ou-se commented:
@ChrisDenton Looks like there's still a left over `#![feature(const_alloc_layout)]` in `library/core/src/lib.rs` causing the CI to fail.

r? @dtolnay for libs fcp

ChrisDenton commented:
Yeah, sorry. In the end I decided to leave in the unstable function because I wasn't sure about stabilizing const on it. If it is stabilized I think it should be const but I was uncertain about committing to that in this PR.

KodrAus commented:
cc @rust-lang/wg-const-eval 

These seem reasonable to me! But let's make sure the const eval group is on board too. This stabilizes a few accessors on `Layout` as const.

@rfcbot fcp merge

oli-obk commented:
> But let's make sure the const eval group is on board too.

These look fine to me

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 0ba3f4c8551df07253cf757d257743abf3559c76 has been approved by `oli-obk`

<!-- @bors r=oli-obk 0ba3f4c8551df07253cf757d257743abf3559c76 -->
<!-- homu: {"type":"Approved","sha":"0ba3f4c8551df07253cf757d257743abf3559c76","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 0ba3f4c8551df07253cf757d257743abf3559c76 with merge a10a301e25729da57aaea6643776088fd794b15d...
<!-- homu: {"type":"BuildStarted","head_sha":"0ba3f4c8551df07253cf757d257743abf3559c76","merge_sha":"a10a301e25729da57aaea6643776088fd794b15d"} -->

memoryruins commented on library/core/src/alloc/layout.rs:
```diff
@@ -50,7 +50,7 @@ impl Layout {
     ///    or equal to `usize::MAX`).
     #[stable(feature = "alloc_layout", since = "1.28.0")]
-    #[rustc_const_unstable(feature = "const_alloc_layout", issue = "67521")]
+    #[rustc_const_stable(feature = "const_alloc_layout", since = "1.49.0")]
```
```suggestion
    #[rustc_const_stable(feature = "const_alloc_layout", since = "1.50.0")]
```
Should this and the others be `1.50.0` now?

RalfJung replied:
Yes, yes it should. Good catch.

ChrisDenton replied:
Ok, done.

RalfJung commented:
@bors r- retry

ChrisDenton committed:
Stabilize `alloc::Layout const` functions

Stabilizes #67521. In particular the following stable methods are stabilized as const fn:

* size
* align
* from_size_align

m-ou-se commented:
@ChrisDenton Looks like you accidentally pulled unrelated commits into this PR with a merge commit. Can you rebase/squash instead?

jyn514 commented:
@m-ou-se we're debugging the issue: https://rust-lang.zulipchat.com/#narrow/stream/122651-general/topic/Git.20help.3F

ChrisDenton commented:
It should be fixed now. Thanks!

m-ou-se commented:
@bors r=oli-obk

bors commented:
:pushpin: Commit 9050d12714e65ed135833b9dfd1b9619c6188ca4 has been approved by `oli-obk`

<!-- @bors r=oli-obk 9050d12714e65ed135833b9dfd1b9619c6188ca4 -->
<!-- homu: {"type":"Approved","sha":"9050d12714e65ed135833b9dfd1b9619c6188ca4","approver":"oli-obk"} -->

pthariensflame commented:
`relnotes` needed?
